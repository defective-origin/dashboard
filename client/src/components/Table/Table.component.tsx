import React from 'react'
import MuiTable from '@mui/material/Table'
import MuiTableContainer from '@mui/material/TableContainer'

// ---| core |---
import { cn } from 'tools'

// ---| pages |---
// ---| screens |---
// ---| components |---
import Scroll from 'components/Scroll'
import Banner from 'components/Banner'

// ---| self |---
import css from './Table.module.scss'
import TableHead from './TableHead'
import TableBody from './TableBody'
import { useTableManager } from './Table.hook'
import TablePagination, { TablePaginationProps } from './TablePagination'
import { TableColumn, TableFilter, TableKeygen, TableRecord } from './Table.type'
import { TableRowMenuItem } from './TableRowMenu'

export type TableProps<T extends TableRecord> = {
  items?: T[]
  width?: number | string
  height?: number | string
  minHeight?: number | string
  columns?: TableColumn<T>[]
  filters?: TableFilter<T>[]
  pagination?: boolean | TablePaginationProps
  loading?: boolean
  actions?: TableRowMenuItem[]
  className?: string
  children?: React.ReactNode
  keygen?: TableKeygen<T>
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Table />
 */
export function Table<T extends TableRecord>(props: TableProps<T>): JSX.Element {
  const { actions, width = '100%', height = '100%', minHeight, loading, items, columns, filters, pagination, keygen, children, className, ...otherProps } = props
  const _className = cn(css.Table, className)
  const manager = useTableManager({ items, columns, filters, pagination, actions })

  // TODO: add Reset button to last column: reset sorting and filters and other column settings
  return (
    <div className={_className} style={{ width, height, minHeight }} {...otherProps}>
      <TablePagination show={!!pagination} {...manager.pagination} />

      <MuiTableContainer className={css.TableContainer} sx={{ height: '100%', width: '100%' }}>
        <MuiTable stickyHeader aria-label='sticky table'>
          <TableHead columns={manager.columns} onSort={manager.sort} />
          <TableBody columns={manager.columns} items={manager.pageItems} keygen={keygen} />

          {children}
        </MuiTable>

        <Scroll key={manager.pageItems.length} v='xy' size='sm' top={60}>
          <Banner
            v='empty'
            loading={loading}
            show={!manager.items.length}
            absolute
          />
        </Scroll>
      </MuiTableContainer>
    </div>
  )
}

Table.displayName = 'Table'

export default Table
