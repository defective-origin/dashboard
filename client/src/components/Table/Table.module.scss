@import 'src/theme';

:root {
  --table-body-filter: drop-shadow(-4px 0 3px rgb(0 0 0 / 10%)) drop-shadow(4px 0 3px rgb(0 0 0 / 10%));
  --table-bg-color: #FFFFFF;
}

.Table {
  border-bottom: 1px solid var(--divider-color);
}
