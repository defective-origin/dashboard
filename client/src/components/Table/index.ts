/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Table.component'

/* INJECT_EXPORT_PLACE */
export * from './Table.component'
export * from './Table.type'
export * from './Table.tool'
