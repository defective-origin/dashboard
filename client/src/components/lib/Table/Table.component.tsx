import React from 'react'
import MuiTable, { TableProps as MuiTableProps } from '@mui/material/Table'

// ---| core |---
// ---| pages |---
// ---| screens |---
// ---| components |---

// ---| common |---
import { cn } from 'common/tools'

// ---| self |---
import css from './Table.module.scss'

export type TableProps = {
  className?: string
  children?: React.ReactNode
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Table />
 */
export function Table(props: TableProps): JSX.Element {
  const { children, className, ...otherProps } = props
  const _className = cn(css.Table, className)

  return <div className={_className} {...otherProps}>{children}</div>
}

Table.displayName = 'Table'

export default Table
