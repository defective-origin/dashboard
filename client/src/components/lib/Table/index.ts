/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Table.component'

/* INJECT_EXPORT_PLACE */
export * from './Table.component'
