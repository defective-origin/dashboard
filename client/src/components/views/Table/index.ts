/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Table.component'

/* INJECT_EXPORT_PLACE */
export * from './Table.component'
export * from './Table.types'
export * from './Table.tools'
export * from './TableRowMenu'
