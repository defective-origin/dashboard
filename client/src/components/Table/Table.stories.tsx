/* eslint-disable no-restricted-imports */
import { field, params } from '../../../.storybook/tool'
import type { Meta, StoryObj } from '@storybook/react'
import Text from 'components/Text'
import Table, { TableProps } from './Table.component'
import { TableColumn } from './Table.type'
import column from './Table.tool'
import { Color } from 'theme'

type Item = {
  name: string;
  code?: string;
  population?: number;
  size?: number;
  density?: number;
  salary?: number;
  gdp?: number;
  date?: Date;
  isBig?: boolean;
}

const COLUMNS: TableColumn<Item>[] = [
  // base config with alignment, data mapping and props providing
  {
    fixed: true,
    name: 'Name',
    cell: Text,
    minWidth: 100,
    align: 'center',
    alignCell: 'left',
    props: { v: 'caption', size: 'xxs', bold: true },
    mapper: (item) => ({ content: item.name, ellipsis: true }),
  },
  column.text({
    field: 'code',
    align: 'center',
    color: 'secondary',
    name: 'ISO\u00a0Code',
    minWidth: 100,
    bold: true,
  }),
  // name taken from field
  column.number({
    field: 'population',
    format: 'amount',
    bold: true,
  }),
  column.icon({
    field: 'isBig',
    name: 'big',
    fill: true,
    mapper: (_, __, is) => is ? { v: 'add' } : { v: 'info' },
  }),
  // fixed column
  column.number({
    fixed: true,
    field: 'size',
    name: 'Size\u00a0(km\u00b2)',
    color: 'warning',
  }),
  column.number({
    name: 'Density',
    field: 'density',
  }),
  column.number({
    name: 'Salary',
    field: 'salary',
    format: 'currency',
  }),
  // manual map data to component props
  column.number({
    fixed: true,
    name: 'GDP',
    field: 'gdp',
    format: 'currency',
    mapper: (_, __, field) => {
      let color: Color = 'primary'

      if (field < 1_000_000) {
        color = 'error'
      } else if (field < 1_000_000_000) {
        color = 'warning'
      } else if (field < 1_000_000_000_000) {
        color = 'success'
      } else {
        color = 'info'
      }

      return { content: field, color }
    },
  }),
  column.date({
    fixed: true,
    field: 'date',
    name: 'Created',
  }),
]

function createData(
  name: string,
  code?: string,
  population?: number,
  size?: number,
  salary?: number,
  gdp?: number,
  date?: Date,
) {
  const hasValue = population && size
  const density = hasValue && population / size
  const isBig = !!size && size > 1_000_000

  return { name, code, population, size, density, salary, gdp, date, isBig }
}

const ITEMS = [
  createData('India', 'IN', 1324171354, 3287263, 2500, 13241713541643, new Date(2000, 1, 1)),
  createData('China', 'CN', 1403500365, 9596961, 1500, 1403500365, new Date(2000, 3, 3)),
  createData('Italy', 'IT', 60483973, 301340, 1700, 60483973163, new Date(2000, 5, 5)),
  createData('United States', 'US', 327167434, 9833520, 2000, 3271674341, new Date(2000, 7, 7)),
  createData('Canada', 'CA', 37602103, 9984670, 2500, 376021031643, new Date(2000, 9, 9)),
  createData('Australia', 'AU', 25475400, 7692024, 3000, 254754006134, new Date(2000, 11, 11)),
  createData('Germany', 'DE', 83019200, 357578, 1700, 830192003, new Date(2000, 2, 15)),
  createData('Ireland', 'IE', 4857000, 70273, 2500, 485700054641, new Date(2000, 4, 15)),
  createData('Mexico', 'MX', 126577691, 1972550, 500, 126577691315, new Date(2000, 5, 15)),
  createData('Japan', 'JP', 126317000, 377973, 1200, 126317000125, new Date(2000, 3, 15)),
  createData('France', 'FR', 67022000, 640679, 3000, 67022000215, new Date(2000, 12, 15)),
  createData('United Kingdom', 'GB', 67545757, 242495, 1000, 67545757124, new Date(2000, 10, 10)),
  createData('Russia', 'RU', 146793744, 17098246, 2900, 146793744421, new Date(2000, 8, 8)),
  createData('Nigeria', 'NG', 200962417, 923768, 2750, 200962, new Date(2000, 6, 6)),
  createData('Brazil', 'BR', 210147125, 8515767, 1970, 210147125521, new Date(2000, 4, 4)),
  createData('Without Value', 'BR', 210147125, 8515767, 1890, 21014712, new Date(2000, 2, 2)),
]

const meta: Meta<typeof Table> = {
  component: Table,
  title: 'Components/Views/Table',
  tags: ['autodocs'],
  argTypes: {
    className: field.string(),
  },
}

export default meta

type Story = StoryObj<React.FC<TableProps<Item>>>

export const Demo: Story = {
  parameters: params('Table'),
  args: {
    columns: COLUMNS,
    items: ITEMS,
    width: 760,
    height: 500,
    pagination: true,
    loading: false,
    actions: [
      { start: 'add', content: 'Add' },
      { start: 'delete', content: 'Delete' },
    ],
    keygen: (item, index) => `${item.name}-${item.code}-${index}`,
  },
}
