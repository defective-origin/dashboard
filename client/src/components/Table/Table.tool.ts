// ---| components |---
import Text, { TextProps } from 'components/Text'

// ---| self |---
import { TableColumn, TableRecord } from './Table.type'

export type TextColumnOptions = Pick<TextProps, 'format' | 'placeholder' | 'color' | 'ellipsis' | 'bold' | 'nowrap'>
export type TextColumn<T extends TableRecord,> = TableColumn<T, TextProps> & TextColumnOptions

export const textColumn = <T extends TableRecord>(column: TextColumn<T>): TextColumn<T> => ({
  sort: true,
  cell: Text,
  mapper: (_, __, field) => ({ content: field }),
  ...column,
  props: {
    v: 'caption',
    size: 'xxs',
    format: column.format,
    placeholder: column.placeholder,
    color: column.color,
    ellipsis: column.ellipsis,
    bold: column.bold,
    nowrap: column.nowrap,
    ...column.props,
  },
})

export const numberColumn = <T extends TableRecord>(column: TextColumn<T>) => textColumn({
  minWidth: 130,
  align: 'right',
  format: 'number',
  ...column,
})

export const dateColumn = <T extends TableRecord>(column: TextColumn<T>) => textColumn({
  minWidth: 200,
  align: 'center',
  format: 'day-of-month-year',
  sort: (_, __, field) => new Date(field),
  ...column,
})

export default {
  text: textColumn,
  number: numberColumn,
  date: dateColumn,
}
