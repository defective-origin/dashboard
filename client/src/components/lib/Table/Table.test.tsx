// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Table from './Table.component'

describe('[Table] component', () => {
  it('should render component', () => {
    const container = render(<Table />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})
