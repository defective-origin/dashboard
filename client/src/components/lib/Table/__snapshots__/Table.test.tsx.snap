// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Table] component > should render component 1`] = `
<DocumentFragment>
  <div
    class="_Table_b2501f"
  />
</DocumentFragment>
`;
